src/NServiceBus.Core/Extensibility/ReadonlyContextBag.cs

```diff
@@ -8,11 +8,19 @@ public interface ReadOnlyContextBag
         /// <summary>
         /// Retrieves the specified type from the context.
         /// </summary>
-        /// <typeparam name="T">The type to retrieve.</typeparam>
+        /// <typeparam name="T">The type to retrieve. The Fullname of will be used to look up the instance.</typeparam>
         /// <returns>The type instance.</returns>
         T Get<T>();
 
         /// <summary>
+        /// Retrieves the specified type from the context.
+        /// </summary>
+        /// <typeparam name="T">The type to retrieve.</typeparam>
+        /// <param name="key">The key of the value being looked up.</param>
+        /// <returns>The type instance.</returns>
+        T Get<T>(string key);
```

Comment:
Needing to change this means we have to wait for a major to do this, because modifying an interface is a breaking change.